group/assignment5/UnknownAlphabet.java

```diff
@@ -0,0 +1,184 @@
+// white = not yet visited; grey = visited, but not yet finished; black = finished.
+class Vertex {
+
+	final char value;
+	Set<Vertex> children = new HashSet<Vertex>();
+	String state = "white";
+
+	Vertex(char value) {
+		this.value = value;
+	}
+
+	public void addChild(Vertex child) {
+		children.add(child);
+	}
+
+	public void setState(String state) {
+		this.state = state;
+	}
+
+	public char getValue() {
+		return value;
+	}
+
+	public Set<Vertex> getChildren() {
+		return children;
+	}
+
+	public String getState() {
+		return state;
+	}
+
+}
+
+// Solution will create the list of vertices and do topological sort to find a valid order.
+// During topological traversal, chars will be saved in reversed order in a stack reversed_alphabet.
+// Used stack to have the correct order using the last-in-first-out property.
+// Keep a list vertices of all the vertices created from the words in dictionary.
+// Keep a map of every char to its vertex to check if a vertex was already created for that char.
+public class UnknownAlphabet {
+
+	Stack<Character> reversed_alphabet = new Stack<Character>();
+	LinkedList<Vertex> vertices = new LinkedList<Vertex>();
+	HashMap<Character, Vertex> char_vertex_map = new HashMap<Character, Vertex>();
+	
+	// Assume dictionary input is a List.
+	public LinkedHashSet<Character> getAlphabet(List<String> dict) throws Exception {
+
+		LinkedHashSet<Character> ordered_alphabet = new LinkedHashSet<Character>();
+
+		if (dict.isEmpty())
+			return ordered_alphabet;
+
+		// Return all characters (no duplicates) if dictionary has only one word.
+		if (dict.size() == 1) {
+			String word = dict.get(0);
+			for (int i = 0; i < word.length(); i++) {
+				ordered_alphabet.add(word.charAt(i));
+			}
+			return ordered_alphabet;
+		}
+
+		createVertices(dict);
+		topologicalSort();
+
+		// Return the characters in the correct order by reversing the stack.
+		while (!reversed_alphabet.isEmpty()) {
+			ordered_alphabet.add(reversed_alphabet.pop());
+		}
+
+		return ordered_alphabet;
+	}
+
+	// Traverse verticies which have not been visited yet.
+	private void topologicalSort() throws Exception {
+
+		for (Vertex vertex : vertices) {
+			if (vertex.getState().equals("white")) {
+				sortFromVertex(vertex);
+			}
+		}
+	}
+
+	// For each unvisisted vertex, visit its children if they haven't been visited (white).
+	// If during the visit of a child we discover it was started but not finished (grey),
+	// then we have a cycle, so no order is possible.
+	private void sortFromVertex(Vertex cur_vertex) throws Exception {
+
+		cur_vertex.setState("grey"); // starting visit, but not finished
+		for (Vertex child : cur_vertex.getChildren()) {
+			if (child.getState().equals("white")) {
+				sortFromVertex(child);
+			} else if (child.getState().equals("grey")) {
+				throw new Exception("There is a cycle, order impossible.");
+			}
+		}
+		cur_vertex.setState("black"); // finished visit.
+		reversed_alphabet.add(cur_vertex.getValue());
```

Comment:
Instead of making a reversed alphabet, I think the solution would work in the same manner if you had a Deque (LinkedList or ArrayDeque) for the alphabet and just add each new character to the front of the queue. This way, you wouldn't have to reverse it once the sorting is done.